hax/helper/generate_sysconf.py

```diff
@@ -0,0 +1,347 @@
+        regex = re.compile('m0conf/profiles/')
+        for key in self.kv_data:
+            match_result = re.match(regex, key['key'])
+            if not match_result:
+                continue
+            return key.split('/')[-1]
+
+
+class ConsulKV:
+    """
+    Helper class to fetch data from consul kv.
+    """
+    def __init__(self, node: str) -> None:
+        self.kv = KVAdapter()
+        self.node = node
+
+    @repeat_if_fails()
+    def get_service_ids(self, svc_name: str) -> List[str]:
+        """
+        Returns the service id by its name assuming that it runs at the
+        current node.
+
+        Examples of the key that will match:
+            m0conf/nodes/srvnode1/processes/6/services/ha
+        """
+        node_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        regex = re.compile(f'^m0conf/nodes/{self.node}\\/processes/([0-9]+)'
+                           f'/services/{svc_name}$')
+        keys = []
+        for key in node_items:
+            match_result = re.match(regex, key['Key'])
+            if not match_result:
+                continue
+            keys.append(match_result.group(1))
+        return keys
+
+    @repeat_if_fails()
+    def get_service_ep(self, proc_id: str) -> Optional[str]:
+        """
+        Returns the service endpoint from the given process id assuming that it
+        runs at the current node.
+
+        Examples of the key that will match:
+            key = m0conf/nodes/srvnode-1/processes/10/endpoint
+            value = inet:tcp:10.230.246.59@3001
+        """
+        key = self.kv.kv_get(
+            f'm0conf/nodes/{self.node}/processes/{proc_id}/endpoint')
+        if key:
+            return key['Value'].decode("utf-8")
+        return None
+
+    @repeat_if_fails()
+    def get_ios_meta_data(self, proc_id: str) -> Optional[str]:
+        """
+        Returns the metadata path for a given ioservice fid.
+
+        Examples of the key that will match:
+            key = m0conf/nodes/srvnode-1.data.private/processes/10/meta_data
+            value = /dev/vg_srvnode-1_md1/lv_raw_md1
+        """
+        key = self.kv.kv_get(
+            f'm0conf/nodes/{self.node}/processes/{proc_id}/meta_data')
+        if key:
+            return key['Value'].decode("utf-8")
+        return None
+
+    @repeat_if_fails()
+    def get_profile_fid(self) -> str:
+        profile_key = self.kv.kv_get('m0conf/profiles/', keys=True)
+        assert profile_key
+        # Get the first profile fid
+        return profile_key[0].split('/')[-1]
+
+
+class Generator:
+    """
+    Generates system configuration files for consul-agent, motr and
+    s3 services.
+    """
+
+    def __init__(self, node: str, kv_file: Optional[str] = None):
+        # Based on the kv_file input, we choose whether to fetch data from
+        # kv file or consul kv.
+        self.provider = KVFile(kv_file=kv_file,
+                               node=node) if kv_file else ConsulKV(node)
```

Comment:
[Optional]
I would add a type hint for `self.provider` (either by making KVFile and ConsulKV to have the same base class or by using [`Union[]`](https://docs.python.org/3/library/typing.html#typing.Union))

Reply:
done.. used Union